PR: Rename variables (by refactoring) with rope

jitseniesen opened the pull request:
I don't think this is quite ready to be merged yet (and given that you are in beta testing, you may prefer to wait till the next release), but I thought it would be good to solicit feedback. I implemented a basic "rename variable" functionality via rope. This is part of issues #415 and #1937.

I'm quite a beginner with Spyder and GitHub and Python GUI so please don't hesitate to comment.
# User experience
1. You put the cursor on the variable that you want to rename:
   ![rename1](https://cloud.githubusercontent.com/assets/7941918/12705008/bee25f40-c85f-11e5-93bb-e6b8256f0f93.png)
2. You select `Rename` from the `Edit` menu:
   ![rename2](https://cloud.githubusercontent.com/assets/7941918/12705016/eb2a98c4-c85f-11e5-84eb-64e0dd4f51b4.png)
3. You type the new name in the variable box:
   ![rename3](https://cloud.githubusercontent.com/assets/7941918/12705023/1c9de9b0-c860-11e5-8506-31b8d123e463.png)
4. And Spyder renames the variable: 
   ![rename4](https://cloud.githubusercontent.com/assets/7941918/12705027/2d573770-c860-11e5-95aa-6cf11cf3b262.png)

In this example, both occurrences of the variable `num` in the function `double` are renamed to `arg`. Note that the variable `num` in the main block is not renamed (which I think is the correct behaviour) even though it is highlighted in yellow at step 1.
# Implementation
1. In `Editor.get_plugin_actions()`, create the new menu item and connect it to `Editor.rename_variable()`.
2. In `Editor.rename_variable()`, find the current editor stack and call its `rename_variable()`.
3. In `EditorStack.rename_variable()`, find the current editor and call its `open_rename_variable_dialog()`.
4. In `CodeEditor.open_rename_variable_dialog()`, find the word currently highlighted and ask the user for the new variable name. Return this information, or `None` if the user cancels.
5. Back in `EditorStack.rename_variable()`, assuming that the user did not cancel, find the associated `PluginManager` and call its `rename()`.
6. `PluginManager.rename()` searches through all registered introspection plugins for one that implements a `rename()` member function and calls it.
7. Only `RopePlugin` has a `rename()` member function. It saves the current buffer contents in a temporary file and uses the rope library to get the new code with the renamed variable.
8. The new code is returned to `EditorStack.rename_variable()` which calls `CodeEditor.change_text()` which replaces the old code with the new code.
# Questions and to do
- Everything is done synchronously. This is easier to implement, but it may also be confusing for the user: otherwise it may be that the user tries to rename a variable, nothing seems to happen for a second so the user continues editing, but then the rename function suddenly returns, the contents of the buffer is replaced with the result, and the user loses the edits. However, is there a possibility of interference with other plugin actions executed in a different thread (I don't check `plugin.busy`)?
- If it is better done asynchronously, I guess it should wait until #2932 is merged.
- Should I have used Qt signals?
- Is it okay to save the contents to a temporary file? The rope library seems to be written with files in mind, but it may be possible to work around it.
- The rename functionality in rope comes with many options; I should check them and maybe expose them to the user. In particular, rope can do a rename over multiple files. Could this work with Spyder projects (which, as I understand it, are going to be re-implemented)? 
- After renaming the variable, the cursor is moved to the top of the file which is probably annoying for the user.
- It needs to be made more robust: what if the user clicks on 'Rename' if the cursor is on Python keyword or anything else that is not an identifier? What if the file being edited is not Python? What if the new name the user types in is not a valid identifier? What if rope is not installed? What if in `RopePlugin`, `project` is `None` (can this happen?)? What if rope throws an exception for any reason? All this needs to be handled gracefully.

ccordoba12 commented:
@jitseniesen, this is very nice!! Thanks for working on it :-)

@Nodd, could you review this one?

Nodd commented on spyderlib/plugins/editor.py:
```diff
@@ -1922,6 +1926,13 @@ def fix_indentation(self):
+    def rename_variable(self):
+        """Open rename variable dialog box."""
+        editorstack = self.get_current_editorstack()
+        if editorstack is not None:
```
Replace `if editorstack is not None:` by `if editorstack:`

Nodd commented on spyderlib/utils/introspection/plugin_manager.py:
```diff
@@ -326,6 +326,24 @@ def show_object_info(self, position, auto=True):
+                continue
+            return fun(code, position, new_name)
+        else:
+            raise NotImplementedError
```
It seems strange to check here if the renaming is possible. If there is no plugin to rename variables, the menu entry should be hidden or disabled. This would also avoid raising `NotImplementedError` in the internal console.

Nodd commented on spyderlib/utils/introspection/rope_plugin.py:
```diff
@@ -239,6 +239,25 @@ def set_pref(self, key, value):
+        self.project.do(changes)
+        result = temp_module.read()
+        temp_module.remove()
+        return result
```
It would be really better to not use a temporary file. I tried to find an alternative but I couldn't find anything quickly so I don't know.
If a temporary file has to be used, it should use a context manager to be sure to remove the file even in case of error, and the name should be more unique that using the pid. See https://docs.python.org/3.5/library/tempfile.html#examples for example.

Nodd commented on spyderlib/widgets/sourcecode/codeeditor.py:
```diff
@@ -915,6 +915,28 @@ def fix_indentation(self):
+
+        Returns (source_code, offset, new_name) on success, None otherwise.
+        """
+        from spyderlib.qt.QtGui import QMessageBox, QInputDialog
```
Imports are always on top of the file !

Nodd commented on spyderlib/widgets/sourcecode/codeeditor.py:
```diff
@@ -915,6 +915,28 @@ def fix_indentation(self):
+            offset = self.get_position('cursor')
+            return (source_code, offset, new_name)
+        else:
+            return None
```
`else: return None` feels strange. Removing these lines would work, but just removing `else` and deindenting the return would be clearer.

Nodd commented:
I had a look at the code but I don't know this part so I don't know if I've missed something.

- I think it's ok to have it synchronous, since all user modifications would be erased
- Where would you use Qt signals ?
- It would be really better to not use temporary files.
- Don't bother with spyder projects, they will be modified and are a bit broken in their current state.

jitseniesen commented:
@Nodd thanks a lot for having a look so promptly and your comments. I agree with all of them. I'll need some time to act on them and to play with rope in order to avoid the temporary files (which I don't like either).

I saw that go-to-definition uses Qt signals to communicate with the introspector so that's why I wondered whether signals should also be used here. However, go-to-definition is asynchronous while rename-variable is synchronous. Besides, it turns out to be actually quite straightforward to implement rename-variable without signals (once I found out how to do it). So I'm happy not to use Qt signals.

Nodd commented:
Yes, it's fine with no signals.

jitseniesen commented:
I addressed @Nodd 's concerns (thanks again), I made some improvements (allow undo, more robust handling of invalid renames), and I rebased (necessary after #2932 which forced me to do some changes). As far as I am concerned, this is ready to be merged.

ccordoba12 commented:
@Nodd, do you have more comments?

Nodd commented:
After a quick overview the code looks fine but I didn't test it.

jitseniesen commented:
I rebased the branch so that it merges cleanly after the changes in pyqt.

ccordoba12 commented:
Thanks a lot @jitseniesen! I'll release beta3 and then I'll come back to you to merge this one :-)

ccordoba12 commented:
@jitseniesen, could you rebase or merge with master to solve conflicts?

@Nodd, could you review this one?

jitseniesen commented:
I will rebase, but I wouldn't mind if this were postponed to 3.1 in the interest of getting 3.0 out of the door as soon as possible.

ccordoba12 commented:
Thanks Jitse!

Nodd commented:
I can have a look next week. I agree that it's not needed for 3.0. We should set a feature freeze at some point.

ccordoba12 commented:
@jitseniesen, please rebase or merge with master so we can review this one for 3.1 :-)

goanpeca commented:
Hi @jitseniesen you are probably in unittest mode, just chiming in to see if you plan to continue with this work :-)

ccordoba12 commented:
Yep, it'd be awesome if you could finish this one and also attach it to our projects :-)

ccordoba12 commented:
This needs to be handled through `pyls` now.
